File: erpnext/utilities/phone_utils.py
import re

import frappe
import phonenumbers
from phonenumbers.phonenumberutil import NumberParseException


def normalize_to_standard_format(phone: str, default_country: str = "VN") -> str:
	if not phone:
		return ""

	# 1. Try parsing with phonenumbers
	try:
		parsed = phonenumbers.parse(phone, default_country)
		if phonenumbers.is_valid_number(parsed):
			formatted = phonenumbers.format_number(parsed, phonenumbers.PhoneNumberFormat.E164)
			return formatted.replace("+", "")
	except NumberParseException:
		pass

	# 2. Fallback logic: extract digits only
	digits = re.sub(r"\D", "", phone)
	if not digits:
		return ""

	# Get country code safely
	try:
		country_code = str(phonenumbers.country_code_for_region(default_country))
	except Exception:
		country_code = "84"

	# Rule A: starts with '0' (e.g., '0901234567' -> '84901234567')
	if digits.startswith("0") and len(digits) >= 9:
		return country_code + digits[1:]

	# Rule B: starts with '840' (e.g., '840901234567' -> '84901234567')
	if digits.startswith(country_code + "0") and len(digits) >= len(country_code) + 9:
		return country_code + digits[len(country_code) + 1 :]

	return digits


def is_valid_phone_number(phone: str, default_country: str = "VN") -> bool:
	if not phone or not str(phone).strip():
		return False
	phone_str = str(phone).strip()
	try:
		parsed = phonenumbers.parse(phone_str, default_country)
		if phonenumbers.is_valid_number(parsed):
			return True
	except Exception:
		pass

	if not phone_str.startswith("+"):
		try:
			parsed = phonenumbers.parse("+" + phone_str)
			if phonenumbers.is_valid_number(parsed):
				return True
		except Exception:
			pass

	return False


def get_phone_variants(phone: str, default_country: str = "VN", for_search: bool = False) -> list:
	"""Generate different formatting variants of a phone number for DB search."""
	if not phone:
		return []

	variants = set([phone])
	digits = re.sub(r"\D", "", phone)
	if digits:
		variants.add(digits)

	if for_search:
		if digits.startswith("84"):
			variants.add("0" + digits[2:])
		elif digits.startswith("0"):
			variants.add("84" + digits[1:])

	try:
		parsed = phonenumbers.parse(phone, default_country)
		if phonenumbers.is_valid_number(parsed):
			formatted = phonenumbers.format_number(parsed, phonenumbers.PhoneNumberFormat.E164)
			variants.add(formatted.replace("+", ""))
			national_format = phonenumbers.format_number(parsed, phonenumbers.PhoneNumberFormat.NATIONAL)
			national_digits = re.sub(r"\D", "", national_format)
			if national_digits:
				variants.add(national_digits)
	except NumberParseException:
		pass

	return list(variants)


def search_doc_by_phone(phone: str, doctype_priority: list = None) -> tuple:
	if not phone:
		return None, None

	if doctype_priority is None:
		doctype_priority = ["Customer", "Lead"]

	variants = get_phone_variants(phone, for_search=True)
	if not variants:
		return None, None

	for doctype in doctype_priority:
		meta = frappe.get_meta(doctype)
		or_filters = {}
		if meta.has_field("normalized_phone"):
			or_filters["normalized_phone"] = ["in", variants]
		if meta.has_field("phone"):
			or_filters["phone"] = ["in", variants]
		if meta.has_field("mobile_no"):
			or_filters["mobile_no"] = ["in", variants]

		if not or_filters:
			continue

		docs = frappe.get_all(doctype, or_filters=or_filters, limit=1, ignore_permissions=True)
		if docs:
			return doctype, docs[0].name

	return None, None

File: erpnext/public/js/utils/phone_utils.js
import { parsePhoneNumber, isValidPhoneNumber, getCountryCallingCode } from 'libphonenumber-js';

frappe.provide("erpnext.utils");

erpnext.utils.normalize_to_standard_format = function(phone, default_country = "VN") {
    if (!phone) return "";

    try {
        const phoneNumber = parsePhoneNumber(phone, default_country);
        if (phoneNumber && phoneNumber.isValid()) {
            let formatted = phoneNumber.format('E.164');
            return formatted.replace("+", "");
        }
    } catch (error) {
        // parsing failed
    }

    let digits = phone.replace(/\D/g, "");
    if (!digits) return "";

    let country_code = "84";
    try {
        country_code = getCountryCallingCode(default_country);
    } catch (e) {}

    if (digits.startsWith("0") && digits.length >= 9) {
        return country_code + digits.substring(1);
    }

    if (digits.startsWith(country_code + "0") && digits.length >= country_code.length + 9) {
        return country_code + digits.substring(country_code.length + 1);
    }

    return digits;
};

erpnext.utils.is_valid_phone_number = function(phone, default_country = "VN") {
    if (!phone || !String(phone).trim()) return false;
    let phone_str = String(phone).trim();
    
    try {
        if (isValidPhoneNumber(phone_str, default_country)) {
            return true;
        }
    } catch (e) {}

    if (!phone_str.startsWith("+")) {
        try {
            if (isValidPhoneNumber("+" + phone_str, default_country)) {
                return true;
            }
        } catch (e) {}
    }

    return false;
};

erpnext.utils.get_phone_variants = function(phone, default_country = "VN") {
    if (!phone) return [];

    let variants = new Set([phone]);
    let digits = phone.replace(/\D/g, "");
    if (digits) {
        variants.add(digits);
    }

    try {
        const phoneNumber = parsePhoneNumber(phone, default_country);
        if (phoneNumber && phoneNumber.isValid()) {
            let formatted = phoneNumber.format('E.164');
            variants.add(formatted.replace("+", ""));
            
            let national_format = phoneNumber.format('NATIONAL');
            let national_digits = national_format.replace(/\D/g, "");
            if (national_digits) {
                variants.add(national_digits);
            }
        }
    } catch (error) {
        // ignore
    }

    return Array.from(variants);
};

erpnext.utils.check_duplicate_phone = function(mobile_no, doctype="Customer") {
    let variants = erpnext.utils.get_phone_variants(mobile_no);

    let or_filters = [];
    
    // Ensure meta is loaded, though for standard doctypes it usually is. 
    // We will just do a safe check. If meta is not loaded, we assume it has both to be safe, 
    // or we can just push only the fields that exist.
    let has_mobile = !frappe.meta.has_field || frappe.meta.has_field(doctype, "mobile_no");
    let has_phone = !frappe.meta.has_field || frappe.meta.has_field(doctype, "phone");

    variants.forEach(v => {
        if (v) {
            if (has_mobile) or_filters.push(["mobile_no", "like", "%" + v + "%"]);
            if (has_phone) or_filters.push(["phone", "like", "%" + v + "%"]);
        }
    });

    if (or_filters.length === 0) {
        frappe.msgprint({
            title: __("Validation Error"),
            message: __("Cannot check duplicate phone: No 'phone' or 'mobile_no' field in Doctype '{0}'", [doctype]),
            indicator: "red"
        });
        return Promise.reject();
    }

    return new Promise((resolve, reject) => {
        frappe.call({
            method: "frappe.client.get_list",
            args: {
                doctype: doctype,
                or_filters: or_filters,
                fields: ["name", doctype === "Customer" ? "customer_name" : "name"]
            },
            callback: function(r) {
                if (r.message && r.message.length > 0) {
                    const existing = r.message[0];
                    let display_name = existing.customer_name || existing.name;
                    frappe.msgprint({
                        title: __("Duplicate Mobile Number"),
                        message: __("Mobile number {0} already exists for {1}: {2}", 
                            [mobile_no, doctype.toLowerCase(), display_name]),
                        indicator: "orange"
                    });
                    reject();
                } else {
                    resolve();
                }
            },
            error: function() {
                reject();
            }
        });
    });
};
